Add support for `--order`, `--orderby` flags in `wp db size` command

kjohnson opened the pull request:
<!--

Thanks for submitting a pull request!

Please review our contributing guidelines if you haven't recently: https://make.wordpress.org/cli/handbook/contributing/#creating-a-pull-request

Here's an overview to our process:

1. One of the project committers will soon provide a code review: https://make.wordpress.org/cli/handbook/code-review/
2. You are expected to address the code review comments in a timely manner (if we don't hear from you in two weeks, we'll consider your pull request abandoned).
3. Please make sure to include functional tests for your changes.
4. The reviewing committer will merge your pull request as soon as it passes code review (and provided it fits within the scope of the project).

You can safely delete this comment.

-->

Resolves https://github.com/wp-cli/db-command/issues/225 by adding support for the `--order` and `--orderby` flags to the `wp db size` command when displaying table sizes.

janw-me commented on features/db-size.feature:
```diff
@@ -202,3 +202,62 @@ Feature: Display database size
+      wp_users	65536 B
+      wp_posts	81920 B
+      wp_comments	98304 B
+      """
```
If WP will ever add/change even 1 option, these numbers will change. And that could happen with every table.
And the order itself might even change.
In all other current tests there is no check with absolute numbers.

My suggestion is after installing, deleting all (sample) posts/post_meta,
Adding a couple of sample posts and postmeta. 
That way the numbers should always be the same.

danielbachhuber replied:
> My suggestion is after installing, deleting all (sample) posts/post_meta,
> Adding a couple of sample posts and postmeta.
> That way the numbers should always be the same.

I think I'd go one step further: don't include absolute numbers at all. I suspect there will still be variables we can't control, which will cause the tests to fail intermittently.

Instead, let's create some contrived scenario where one table obviously has more data than another, and then verify the expected behavior with our `--order=<order>` argument.

kjohnson replied:
I had originally pulled the hard-coded sizes based on another of the `size.feature` tests:
```feature
    When I run `wp db size --tables`
    Then STDOUT should contain:
      """
      wp_posts	81920 B
      """
```

I'll look into how this could be stabilized without relying on hard-coded values and specific orders.

danielbachhuber replied:
@kjohnson Oh! You could read the current table size to an environment variable, and then use that in your test.

kjohnson replied:
I figured out that if I switch to a JSON output format, then I can more easily determine the first row based on the JSON syntax.

```feature
  Scenario: Display ordered table sizes for a WordPress install
    Given a WP install
    And I run `wp post generate --post_type=page --post_status=draft --count=300`

    When I run `wp db size --tables --order=desc --orderby=size --format=json`
    Then STDOUT should contain:
      """
      [{"Name":"wp_posts",
      """
```

danielbachhuber commented on src/DB_Command.php:
```diff
@@ -904,6 +904,24 @@ public function tables( $args, $assoc_args ) {
+	 * default: ASC
+	 * options:
+	 *   - ASC
+	 *   - DESC
```
I think my preference is for `asc` and `desc`. `wp comment meta list` and `wp option list` both use lowercase ([ref](https://developer.wordpress.org/cli/commands/comment/meta/list/), [ref](https://developer.wordpress.org/cli/commands/option/list/)), so I think there's sufficient prior art there.

kjohnson replied:
When I searched for `--order` the first example that came up was form the profile command, which uses uppercase 😅 

I see now that the consensus is lowercase.

kjohnson replied:
Resolved by https://github.com/wp-cli/db-command/pull/226/commits/621ee4bc976698fd348ab510959557b8c7f73d2f

danielbachhuber commented on src/DB_Command.php:
```diff
@@ -1007,6 +1030,7 @@ public function size( $args, $assoc_args ) {
 			$rows[] = [
 				'Name' => DB_NAME,
 				'Size' => strtoupper( $db_bytes ) . $default_unit,
+				'_Bytes' => strtoupper( $db_bytes ),
```
Is there a reason we need to add an underscore to `_Bytes`? I think it's fine to include in the output...

kjohnson replied:
The fields are configured above that line to only output `Name` and `Size`, so I added an `_` to the `Bytes` temporary key as a convention for visibility. It doesn't change the code either way, so it can be easily removed if you'd like.

danielbachhuber replied:
Cool, let's go ahead and remove the underscore. Thanks!

kjohnson replied:
Resolved by https://github.com/wp-cli/db-command/pull/226/commits/8ba0f6aae72e2d1bf17e147c727935478ef2aab3.

danielbachhuber reviewed (changes requested):
Great start, @kjohnson !

Overall, it looks really good. Left a few comments with things to be cleaned up. Once they're done, we can 🚢

danielbachhuber commented on src/DB_Command.php:
```diff
@@ -953,6 +970,8 @@ public function size( $args, $assoc_args ) {
 		$tables                 = ! empty( $tables );
 		$all_tables             = Utils\get_flag_value( $assoc_args, 'all-tables' );
 		$all_tables_with_prefix = Utils\get_flag_value( $assoc_args, 'all-tables-with-prefix' );
+		$order                  = Utils\get_flag_value( $assoc_args, 'order', 'ASC' );
```
```suggestion
		$order                  = Utils\get_flag_value( $assoc_args, 'order', 'asc' );
```

danielbachhuber reviewed (approved):
Nice work on this, @kjohnson !

danielbachhuber commented:
@kjohnson Any ideas why that test is failing? ^

kjohnson commented:
@danielbachhuber intermittently it looks like the posts are not being generated. When I run `wp post list --post_type=page --format=count` is sometimes comes back as `1`.

```feature
    And I run `wp post generate --post_type=page --post_status=draft --count=300`
    And I run `wp post list --post_type=page --format=count`
    Then STDOUT should be:
    """
    300
    """
```

kjohnson commented:
I've added and removed `wp site empty --yes` during local testing as I tried to figure out how to test this. I just added it back which might help with the above failed tests.

kjohnson commented:
Funny that the DB Command is having database related issues during testing. 😅 

Are we able to re-run tests on `master` to confirm that this issue is related to this PR?

danielbachhuber commented:
> Are we able to re-run tests on `master` to confirm that this issue is related to this PR?

Looks like there's a WordPress trunk failure on `master`:

<img width="799" alt="image" src="https://user-images.githubusercontent.com/36432/192606934-da8f81ca-a450-4bfb-961f-5c6be8607234.png">

danielbachhuber commented:
> Looks like there's a WordPress trunk failure on `master`:

Here's the issue: https://core.trac.wordpress.org/ticket/56617

kjohnson commented:
Ok, yeah, looks like a day earlier and this would have been in the clear 😅 

Nice to see that there is a patch ready to go.

danielbachhuber commented:
These are pretty annoying test failures!

I spent some time debugging again today. As far as I can tell, there might be a couple of issues:
1. The table size is cached in `information_schema.TABLES`. Sometimes the cache updates, sometimes it doesn't. [This might be a long-term issue](https://stackoverflow.com/a/64214609). I wasn't able to get `FLUSH TABLES` or `ANALYZE TABLE %s` to reliably update he cache.
2. Empty table sizes may actually vary based on storage engine. 

I've manually verified your new arguments worked, and marked the tests `@broken` for now: 4f331be1c4e043e841c9d6b5e1ebd92b8423e88f

Appreciate your efforts on this!
